### ISSUE TITLE
SSL connection fails on DNS round robin and ssl_context.check_hostname = True

### ISSUE THREAD

**Issue by petervanhooft** [vertica/vertica-python#257]
We use a DNS round robin to connect to a Vertica cluster. Connecting with SSL and ssl_context.check_hostname = True doesn't work, because ssl_options.wrap_socket(raw_socket, server_hostname=host) in enable_ssl gets an IP address passed as server_hostname parameter, and that won't work with certificate hostname check.
Proposed patch:

[patch.txt](https://github.com/vertica/vertica-python/files/2832327/patch.txt)

peter

**Comment by sitingren** [comment]
@petervanhooft Thanks for reporting this bug! The test suite lacks of SSL testing, so it is hard to verify it works.

Here are some of my advise to your patch:
```diff
--- vertica/connection.py-org	2019-02-05 14:20:49.345768826 +0100
+++ vertica/connection.py	2019-02-05 14:14:39.833451310 +0100
@@ -308,7 +308,7 @@
         ssl_options = self.options.get('ssl')
         self._logger.debug('SSL option is {0}'.format('enabled' if ssl_options else 'disabled'))
         if ssl_options:
-            raw_socket = self.enable_ssl(raw_socket, ssl_options)
+            raw_socket = self.enable_ssl(raw_socket, ssl_options, self.options['host'])
+ # "self.options['host']" would only work in specific situations.
+ # Since we introduced connection load balancing and failover (i.e. 'backup_server_node' option in
+ # vertica_python.connect()), which happens right before enabling SSL, so the server hostname may
+ # not be self.options['host'], but a hostname in self.options['backup_server_node']. You may need
+ # to save the actual connected hostname to a variable.
+ # I'm not super familiar with SSL, but is it reasonable to get the hostname from IP address with
+ # `socket.gethostbyaddr(host_ip)`?
 
         self.socket = raw_socket
         return self.socket
@@ -354,7 +354,7 @@
 
         return raw_socket
 
-    def enable_ssl(self, raw_socket, ssl_options):
+    def enable_ssl(self, raw_socket, ssl_options, server_hostname):
         from ssl import CertificateError, SSLError
         # Send SSL request and read server response
         raw_socket.sendall(messages.SslRequest().get_message())
@@ -364,7 +364,7 @@
             try:
                 if isinstance(ssl_options, ssl.SSLContext):
-                    host, port = raw_socket.getpeername()  <= This line would be useless if you pass a server_hostname value
-                    raw_socket = ssl_options.wrap_socket(raw_socket, server_hostname=host)
+                    raw_socket = ssl_options.wrap_socket(raw_socket, server_hostname=server_hostname)
                 else:
                     raw_socket = ssl.wrap_socket(raw_socket)
             except CertificateError as e:
```
It'd be great if you can share your settings and an example program for reproducing this bug.

You can create a pull request once you test your bugfix manually (Testing with different python versions would be better!). Please see [CONTRIBUTING.md](https://github.com/vertica/vertica-python/blob/master/CONTRIBUTING.md) for how to create a pull request. We can postpone adding SSL connection tests to the test suite now. Thank you again!

**Comment by petervanhooft** [comment]
Hi, I'm told by legal that I can't sign the CLA, but I can submit the patch directly.
- is it reasonable to get the hostname from IP address with`socket.gethostbyaddr(host_ip)`?
No, because you can have only one certificate per Vertica cluster, so the hostname won't match the certificate. I didn't test the case yet where you have your vertica cluster in a subdomain and have a wildcard certificate.

The patch I attached works with load balancing on/off, ssl on/off and ssl_context.check_hostname on/off and passes all tox tests.
[vertica-ssl-patch.txt](https://github.com/vertica/vertica-python/files/2900113/vertica-ssl-patch.txt)

peter

**Comment by sitingren** [comment]
@petervanhooft Ok, I'm sorry that you cannot sign the CLA. We are glad to listen if you would like to tell us the reason, so that we can see if there is a workaround. If you don't sign the CLA, you are unable to merge your work to the repo. Do you authorize me to create a new PR with your patch and merge to master branch?

**Comment by petervanhooft** [comment]
Hi, this is what I received:
I reviewed the CLA and I do not like the way it is signed only once. Once signed, it does not only cover your current patch but any future submission - I assume that any future submission bypasses the CLA (until the CLA is changed) and that such gating is done on a per-company and not on a per-user basis - in that case we have no control over what we are submitting in the future.

You are authorized to create a new PR with this patch and merge to master branch.

peter

**Comment by sitingren** [comment]
@petervanhooft The lawyer from our side is still thinking about the CLA issue; in the meantime we need more test framework work to support SSL anyway, so we will likely want to implement your change ourselves as part of a broader effort to better support SSL. I'll keep this issue open until we decide what we will do. Thanks!

**Comment by dmitrypolo** [comment]
@sitingren is there any update on this issue, the company I work at currently faces the same issue. We have load balancing and having the `check_hostname` flag set to `True` causes failure. It sounds like you want to add more `SSL` tests, perhaps you can elaborate what you are looking for and I can try to get a PR together.

**Comment by sitingren** [comment]
Thanks @dmitrypolo, you are welcome to create a new PR.

Adding SSL tests requires server configuration and client cert setup, which means we need to modify the test framework (.travis.yml and maybe other files). Currently I don't have a good idea about how hard it is.

If you don't add SSL tests with your PR, I recommend testing your bugfix manually with all combinations like #258. As of today, our target python versions are 2.7, 3.5, 3.6, 3.7.
